src/LogicTasks/Syntax/ComposeFormula.hs

```diff
@@ -0,0 +1,167 @@
+{-# LANGUAGE ApplicativeDo #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE RecordWildCards #-}
+
+module LogicTasks.Syntax.ComposeFormula where
+
+
+import Control.Monad.IO.Class(MonadIO (liftIO))
+import Control.Monad.Output (
+  GenericOutputMonad (..),
+  LangM,
+  OutputMonad,
+  ($=<<),
+  english,
+  german, translate, localise, translations,
+  )
+import Data.Maybe (fromJust, isNothing, isJust)
+
+import LogicTasks.Helpers (extra, fullKey, instruct, keyHeading, reject, example)
+import Tasks.SynTree.Config (checkSynTreeConfig, SynTreeConfig)
+import Trees.Types (TreeFormulaAnswer(..), SynTree (Binary), showOperator)
+import Control.Monad (when)
+import Trees.Print (transferToPicture, display)
+import Tasks.ComposeFormula.Config (ComposeFormulaInst(..))
+import Trees.Helpers (collectLeaves, collectUniqueBinOpsInSynTree)
+import Data.Containers.ListUtils (nubOrd)
+import LogicTasks.Syntax.TreeToFormula (cacheTree)
+import Data.Foldable (for_)
+
+
+
+
+description :: (OutputMonad m, MonadIO m) => FilePath -> ComposeFormulaInst -> LangM m
+description path ComposeFormulaInst{..} = do
+    instruct $ do
+      english $ "Imagine that the two displayed " ++ eTreesOrFormulas ++ " are hung below a root node with operator "
+      english $ showOperator operator
+      english $ ". One " ++ eTreeOrFormula ++" on the left and the other " ++ eTreeOrFormula ++" on the right, and once the other way round."
+      german $ "Stellen Sie sich vor, die beiden angezeigten " ++ gTreesOrFormulas ++" würden unterhalb eines Wurzelknotens mit Operator "
+      german $ showOperator operator
+      german $ " gehängt. Einmal " ++ derDie ++" eine " ++ gTreeOrFormula ++" links und " ++ derDie ++" andere " ++ gTreeOrFormula ++" rechts, und einmal genau andersherum."
+
+    when (isJust leftTreeImage) $ image $=<< liftIO $ cacheTree (fromJust leftTreeImage) path
+    when (isNothing leftTreeImage) $ paragraph $ code $ display leftTree
+
+    when (isJust rightTreeImage) $ image $=<< liftIO $ cacheTree (fromJust rightTreeImage) path
+    when (isNothing rightTreeImage) $ paragraph $ code $ display rightTree
```

Comment:
Hier wäre sicher jeweils ein `case leftTreeImage of` und `case rightTreeImage of` besser als die Verwendungen von `isJust/fromJust/isNothing`.